### DISCUSSION TITLE
✨ API for export of `Directory` and `File` from `Service` container

### DISCUSSION THREAD

**Discussion by jpadams** [dagger/dagger#12353] [Ideas]
### What are you trying to do?

Users have asked for the ability to get files out of Service containers so they can capture test reports, or logs, or other generated files. Currently there is no API.

> integration test container cannot be a service if we want to extract files like tests reports at the end of a run. If it was possible to extract files from a service that would also work.

#### Discord references:
https://discord.com/channels/707636530424053791/1239817495309713428/1239817495309713428
 from @sipsma:
 > In the past it was a technical limitation, but something I'm coincidentally working on at this exact moment will lift that limitation, so yeah I think it would make sense at this point.
>
> It would still currently be required that the Service is still running, which makes it a bit tricky for the case of automatically started/stopped services, but we have support for manual start/stop now too so it would make sense there. I think this would have to be in the bucket of "advanced use cases" since the fact that the service must still be running is a requirement on the user, not just us, but that's probably okay to start out with.
cc @vito I think the stuff I'm writing to enable serving nested sessions from the executor could be re-used here, specifically the filesync part of it

https://discord.com/channels/707636530424053791/1164508519588958348/1165002974926487632
<img width="600" alt="image" src="https://github.com/dagger/dagger/assets/3187222/f25b59e0-85db-4961-8eba-a862d604d567">

https://discord.com/channels/707636530424053791/1180660724452818985/1181944244345258018
<img width="600" alt="image" src="https://github.com/dagger/dagger/assets/3187222/4c5e1707-36e1-4685-83f7-d12d713b52d8">

### Why is this important to you?

This unlocks some advanced, real-world test use-cases and is expected since a `Service` container is also a `Container` in some sense and the API to access files and directories is expected to work.

### How are you currently working around this?

There may be workarounds with volumes, but nothing first class.

**Comment by jedevc** [comment]
Probably worth x-linking this to https://github.com/dagger/dagger/issues/6553 - which is the same, but for `Stdout`/`Stderr` and `ExitCode`.
